### PR TITLE
Remove synthetic, private, protected and static methods

### LINKED ISSUE #288: Private, protected and static methods accidentally exposed
```clojure
> (-> #'amazonica.aws.s3/add-acl-headers meta :amazonica/methods first .getModifiers)
10
```

That means "private static". You can check this thusly:

```clojure
> (bit-or java.lang.reflect.Modifier/PRIVATE java.lang.reflect.Modifier/STATIC)
10
```

I'm guessing that neither private nor static methods were intended to be exposed as client methods. It seems obvious that all exposed fns should map to public instance methods. 

I found the following private methods exposed as fns:

```clojure
(defn is-private-method? [fn-var] (-> fn-var meta :amazonica/methods first .getModifiers java.lang.reflect.Modifier/isPrivate))
;; (filter is-private-method? all-aws-fns) =>
(#'amazonica.aws.cloudsearchdomain/anonymous-invoke
 #'amazonica.aws.cloudsearchdomain/do-invoke
 #'amazonica.aws.cloudsearchdomain/init
 #'amazonica.aws.cloudsearchdomain/invoke
 #'amazonica.aws.lambda/anonymous-invoke
 #'amazonica.aws.lambda/do-invoke
 #'amazonica.aws.lambda/init
 #'amazonica.aws.lambda/invoke
 #'amazonica.aws.s3/add-acl-headers
 #'amazonica.aws.s3/add-date-header
 #'amazonica.aws.s3/add-header-if-not-null
 #'amazonica.aws.s3/add-parameter-if-not-null
 #'amazonica.aws.s3/add-part-number-if-not-null
 #'amazonica.aws.s3/add-response-header-parameters
 #'amazonica.aws.s3/add-string-list-header
 #'amazonica.aws.s3/are-implicit-global-clients-disabled
 #'amazonica.aws.s3/assert-https
 #'amazonica.aws.s3/before-request
 #'amazonica.aws.s3/bucket-region-should-be-cached
 #'amazonica.aws.s3/build-default-endpoint-resolver
 #'amazonica.aws.s3/build-endpoint-resolver
 #'amazonica.aws.s3/check-https
 #'amazonica.aws.s3/create-put-object-result
 #'amazonica.aws.s3/create-sig-v2-signer
 #'amazonica.aws.s3/do-upload-part
 #'amazonica.aws.s3/fetch-region-from-cache
 #'amazonica.aws.s3/get-acl
 #'amazonica.aws.s3/get-bucket-region-via-head-request
 #'amazonica.aws.s3/get-bucket-request-payment
 #'amazonica.aws.s3/get-builder
 #'amazonica.aws.s3/get-create-bucket-endpoint
 #'amazonica.aws.s3/get-protocol
 #'amazonica.aws.s3/get-signer-region
 #'amazonica.aws.s3/is-signer-overridden
 #'amazonica.aws.s3/is-standard-endpoint
 #'amazonica.aws.s3/no-explicit-region-provided
 #'amazonica.aws.s3/populate-request-with-copy-object-parameters
 #'amazonica.aws.s3/populate-request-with-copy-part-parameters
 #'amazonica.aws.s3/populate-request-with-mfa-details
 #'amazonica.aws.s3/populate-source-sse_-c
 #'amazonica.aws.s3/populate-sse_-c
 #'amazonica.aws.s3/populate-sse_-kms
 #'amazonica.aws.s3/put-local-object
 #'amazonica.aws.s3/reject-null
 #'amazonica.aws.s3/resolve-request-endpoint
 #'amazonica.aws.s3/set-acl
 #'amazonica.aws.s3/set-bucket-request-payment
 #'amazonica.aws.s3/set-content
 #'amazonica.aws.s3/set-zero-content-length
 #'amazonica.aws.s3/should-perform-head-request-to-find-region
 #'amazonica.aws.s3/should-retry-complete-multipart-upload
 #'amazonica.aws.s3/update-sig-v4-signer-with-region
 #'amazonica.aws.s3/url-encode-tags
 #'amazonica.aws.s3/validate-request-before-transmit
 #'amazonica.aws.s3transfer/assert-parameter-not-null
 #'amazonica.aws.s3transfer/check-mutability
 #'amazonica.aws.s3transfer/do-download
 #'amazonica.aws.s3transfer/do-upload
 #'amazonica.aws.s3transfer/is-s3-object-modified-since-pause
 #'amazonica.aws.s3transfer/list-files
 #'amazonica.aws.s3transfer/shutdown-thread-pools)
```

These mostly look like garbage to me. Note presence of lambda and cloudsearchdomains which probably answers my question on #287. Is removing those two clients explicitly just a workaround for the fact that you care about some of those methods but _only_ for those two services?

The following static methods are exposed:

```clojure
> (filter is-static-method? all-aws-fns)
(#'amazonica.aws.s3/add-acl-headers
 #'amazonica.aws.s3/add-date-header
 #'amazonica.aws.s3/add-header-if-not-null
 #'amazonica.aws.s3/add-parameter-if-not-null
 #'amazonica.aws.s3/add-response-header-parameters
 #'amazonica.aws.s3/add-string-list-header
 #'amazonica.aws.s3/create-put-object-result
 #'amazonica.aws.s3/get-bucket-region-cache
 #'amazonica.aws.s3/get-protocol
 #'amazonica.aws.s3/populate-request-metadata
 #'amazonica.aws.s3/populate-request-with-copy-part-parameters
 #'amazonica.aws.s3/populate-requester-pays-header
 #'amazonica.aws.s3/populate-source-sse_-c
 #'amazonica.aws.s3/populate-sse_-c
 #'amazonica.aws.s3/populate-sse_-kms
 #'amazonica.aws.s3transfer/append-multipart-user-agent
 #'amazonica.aws.s3transfer/append-single-object-user-agent)
```

There is some overlap, but not full overlap:

```
> (filter (every-pred (complement is-private-method?) is-static-method?) all-aws-fns)
(#'amazonica.aws.s3/get-bucket-region-cache
 #'amazonica.aws.s3/populate-request-metadata
 #'amazonica.aws.s3/populate-requester-pays-header
 #'amazonica.aws.s3transfer/append-multipart-user-agent
 #'amazonica.aws.s3transfer/append-single-object-user-agent)
```

`getBucketRegionCache` has package-level visibility, so it is certainly not intended to be exposed. The following protected methods are exposed:

```clojure
;;; take a wild guess what the implementation for is-protected-method? looks like ;)
> (filter is-protected-method? all-aws-fns)
(#'amazonica.aws.dynamodbv2/calculate-crc32-from-compressed-data
 #'amazonica.aws.s3/create-request
 #'amazonica.aws.s3/create-signer
 #'amazonica.aws.s3/create-signer-provider
 #'amazonica.aws.s3/new-initiate-multipart-upload-request
 #'amazonica.aws.s3/populate-request-metadata
 #'amazonica.aws.s3/populate-requester-pays-header
 #'amazonica.aws.s3/presign-request
 #'amazonica.aws.s3/use-strict-host-name-verification
 #'amazonica.aws.s3transfer/finalize
 #'amazonica.aws.sqs/create-signer-provider)
```

These also look like garbage to me :)

I'd be happy to add a fix for this, but #286 should be fixed by merging #287 first since that puts us in a much better position to address the issue. I think the fix is "filter private methods". It is unclear what to do about cloudsearchdomains and lambda, but I'm guessing you remember why that is :)

### PR DESCRIPTION
**Pull request by lvh**
Closes #288.

### CODE REVIEW

**Comment by mcohen01** [comment]
👍

**Comment by lvh** [comment]
I found a place where the code was already filtering for private methods, so I got rid of that since it's no longer necessary.

**Comment by lvh** [comment]
(You should probably sanity check that before I click merge :D)

File: src/amazonica/core.clj [hunk 1]
```diff
@@ -897,8 +897,7 @@
   "Finds the appropriate method to invoke in cases where
   the Amazon*Client has overloaded methods by arity or type."
   [methods & arg]
-  (let [args (:args (args-from arg))
-        methods (filter #(not (Modifier/isPrivate (.getModifiers %))) methods)]
+  (let [args (:args (args-from arg))]
```
Comment (mcohen01) [inline src/amazonica/core.clj]: Nice catch!

**Comment by lvh** [comment]
I'm not sure about the process here. Is it OK if I merge this PR myself?

**Comment by mcohen01** [comment]
There is no process, merge away.

**Comment by lvh** [comment]
BAM

**Pull request merged**